src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java

```diff
@@ -64,13 +67,28 @@ public AbstractTaskActor() {
 	 * @throws NullPointerException on case of no match
 	 */
 	protected Resource getFirstResourceMatchingType(Model request, Resource type) {
-		ResIterator matches = request.listResourcesWithProperty(RDF.type, type);
+		//taken from https://jena.apache.org/documentation/inference/#RDFSPlusRules
+		//TODO: read from local file: https://github.com/Multi-User-Domain/mud-jena/issues/9
+		List<Rule> rules = Rule.rulesFromURL("https://calum.inrupt.net/public/rules/rdfsbasicrules.txt");
+		
+		//TODO: https://github.com/Multi-User-Domain/mud-jena/issues/14
+		Model buildings = ModelFactory.createDefaultModel();
+		buildings.read(MUDBuildings.getURI(), null, "TTL");
+		buildings.read(MUD.getURI(), null, "TTL");
+		
+		//we are performing transitive reasoning: if A is type B and B is subclass of C, then A is C as well
+		GenericRuleReasoner reasoner = new GenericRuleReasoner(rules);
+		reasoner.setOWLTranslation(true);               // not needed in RDFS case
+		reasoner.setTransitiveClosureCaching(true);
+		InfModel inf = ModelFactory.createInfModel(reasoner, buildings, request);
+		
+		ResIterator matches = inf.listResourcesWithProperty(RDF.type, type);
 		try {
 			Resource match = matches.next();
 			return match;
 		}
 		catch(java.util.NoSuchElementException e) {
-			throw new NullPointerException("An object with type " + type + " must be passed with the request");
+			throw new MissingRequiredArgumentException("An object with type " + type + " must be passed with the request");
```

Comment:
Can we achieve the same functionality without the mappers and custom types by throwing a `BadRequestException` here instead?

https://docs.oracle.com/javaee/7/api/javax/ws/rs/BadRequestException.html

Reply:
Yes we could :grin: good find, thanks!